### DISCUSSION TITLE
Keybinding/command to change file sorting method on-the-fly

### DISCUSSION THREAD

**Discussion by keithfancher** [nvim-tree/nvim-tree.lua#1922] [Ideas]
I'd love to be able to, e.g., quickly switch to a "recently modified" sort at times, but don't want that to always be the default.

Is this currently doable? Not sure if it's possible (or even desirable) to call `setup()` multiple times with different params via a keybinding, or if there would be an easier way.

Thanks!

**Comment by alex-courtis** [comment]
You can customise the sorting with your own function, see `:help nvim-tree.sort_by`.

You can define a mapping to flip between your desired sort configrurations. Here's a complete example: https://github.com/nvim-tree/nvim-tree.lua/pull/1566#issuecomment-1320737448

We should add a recipe for this to the wiki. I'd be most grateful if you could do :)

**Reply by keithfancher** [reply]
Thanks for the quick reply!

> You can define a mapping to flip between your desired sort configrurations. Here's a complete example: https://github.com/nvim-tree/nvim-tree.lua/pull/1566#issuecomment-1320737448

Ah, cool! But it doesn't seem like "modification time" is available to the custom sorting function... right? Without that, the only way to sort by `"modification_time"` _appears_ to be to call `setup` again... or do some gymnastics with some internal bits that we probably shouldn't :D

For example, just as a quick hack to get the functionality I needed, I stuck the following two lines in my `init.vim`:

```
" Probably cleaner to actually stick these in lua functions and keymap in `setup`? This works, though!
noremap <leader>tr :lua require("nvim-tree.explorer.sorters").setup({sort_by="modification_time"})<CR>:lua require("nvim-tree.api").tree.reload()<CR>
noremap <leader>tn :lua require("nvim-tree.explorer.sorters").setup({sort_by="name"})<CR>:lua require("nvim-tree.api").tree.reload()<CR>
```

Works great as a temporary solution, but perhaps the long-term solution is to expose a function to change the sort method (and refresh) via the API? For my use-case at least, I don't need to track state or anything fancy... just binding different keys to different sort methods is plenty!

(And defining custom sort functions is also overkill for me -- I just want to change between a couple of the provided existing methods.)

**Reply by alex-courtis** [reply]
> Works great as a temporary solution, but perhaps the long-term solution is to expose a function to change the sort method (and refresh) via the API? For my use-case at least, I don't need to track state or anything fancy... just binding different keys to different sort methods is plenty!
> 
> (And defining custom sort functions is also overkill for me -- I just want to change between a couple of the provided existing methods.)

How about something like this:

```lua
local my_sort_by_method = "name"

local my_sort_by_function = function(nodes)
  if my_sort_by_method == "name" then
    my_sort_by_method = "modification_time"
  else
    my_sort_by_method = "name"
  end

  return my_sort_by_method
end
```

Returning the method string would need to be built.

**Reply by alex-courtis** [reply]
Just new API:

```lua
api.tree.sort_by("modification_time")
```

**Reply by alex-courtis** [reply]
I'll get back to you.

**Reply by keithfancher** [reply]
Cool! Yeah, I think just exposing the `sort_by` via the API (for reads and writes) would let folks customize it to their heart's content.

I think this would also solve @petobens request in [this comment](https://github.com/nvim-tree/nvim-tree.lua/pull/1566#issuecomment-1328060354), if I understand it correctly. And I'm sure others would appreciate it as well!

**Reply by petobens** [reply]
Yep this exactly the same thing I would like to see in the api :)

**Comment by alex-courtis** [comment]
Please see [Recipe: Cycle Sort Methods](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#cycle-sort-methods)

**Comment by va9iff** [comment]
doesn't it make sense to have something like `api.set_sorter( string | function )` ? it's unnecessarily crumblesome to define a separate variable to keep track of the current sorting method and a separate function for just to return that variable, only to pass the string `"modified_time"` or `"name"` to nvim-tree which can be done like `api.set_sorter("modified_time")` or `api.set_sorter("name")` so, maybe consider simplifying it to that

**Comment by va9iff** [comment]
```lua
local sorter_string = "name"
local sorter_function = function(nodes) 
	return sorter_string 
end

local sort_by_time = function ()
	sorter_string = "modification_time"
	require('nvim-tree.api').tree.reload()
end

local sort_by_name = function ()
	sorter_string = "name"
	require('nvim-tree.api').tree.reload()
end

require("nvim-tree").setup({ 
	sort = { sorter = sorter_function } 
})
```
just passing a string should not require that much boilerplate imo
